language-support/ts/codegen/tests/ts/build-and-lint-test/src/__tests__/test.ts

```diff
@@ -574,6 +574,48 @@ describe("interface definition", () => {
       templateId: emptyIfcId,
     });
   });
+  describe("choice name collision", () => {
+    // statically assert that an expression is a choice
+    const theChoice = <T extends object, C, R, K>(c: Choice<T, C, R, K>) => c;
+
+    // Something is inherited
+    test("unambiguous inherited is inherited", () => {
```

Comment:
There are two reasons to do this.

First, it is more compatible with the prior implementation. We had reason to completely remove the feature in Java/Scala, because the exercise implementation change. That didn't apply here, since choices already carry their template/interface IDs.

Second, as I mention in https://github.com/digital-asset/daml/issues/13668#issuecomment-1110146425, invoking a choice via concrete template is currently the only type-safe way. Java and Scala have the benefit of static guides for their contract ID conversion `toInterface`: only implemented interface targets compile. TypeScript currently does not. I would prefer users navigate uncharted waters only when it is truly necessary (i.e. to solve this ambiguity).

Reply:
> First, it is more compatible with the prior implementation. We had reason to completely remove the feature in Java/Scala, because the exercise implementation change. That didn't apply here, since choices already carry their template/interface IDs.

I don’t think this is a strong argument. Interfaces are in early access. Now is the time to break things if we think it’s an improvement. I find it super unintuitive that if I have one retroactive instance and one normal instance, the "normal" one will silently be used without telling me about the ambiguity.

> Second, as I mention in https://github.com/digital-asset/daml/issues/13668#issuecomment-1110146425, invoking a choice via concrete template is currently the only type-safe way. Java and Scala have the benefit of static guides for their contract ID conversion toInterface: only implemented interface targets compile. TypeScript currently does not. I would prefer users navigate uncharted waters only when it is truly necessary (i.e. to solve this ambiguity).

Do we have some way to make that typesafe? The way we’re currently imagining our usage we’ll use retroactive instances pretty heavily so no form of typesafety for that would be pretty sad even if not a dealbreaker.

Reply:
> I find it super unintuitive that if I have one retroactive instance and one normal instance, the "normal" one will silently be used without telling me about the ambiguity.

Why do you expect that to happen? There is nothing about #14082 that would suggest that retroactive instances would be treated any differently from "normal" instances in codegen output, since codegen operates on whole environments, not individual packages. It will be a normal ambiguous no-inheritance case.

> Do we have some way to make that typesafe?

I don't know, but as the design comment mentioned, I consider trying to solve that problem out of scope for this PR.

Reply:
> Now I add another DAR which defines a choice with the same name and then it no longer compiles

I get that; @stefanobaghino-da and @realvictorprm similarly advocate for "add only adds", hence the alternative design in Java codegen. I'm more comfortable with not preserving this property, I suppose, because in programming languages broadly, [adding overloads often breaks callers of the existing ones](https://discuss.daml.com/t/scala-inference-rules-of-thumb/2242#method-overloading-breaks-all-kinds-of-inference-8).

I suggest we try to fix the type-safety of interface-sourced choice exercise in a separate issue, and then reconsider unambiguous inheritance.

Reply:
> in programming languages broadly, [adding overloads often breaks callers of the existing ones](https://discuss.daml.com/t/scala-inference-rules-of-thumb/2242#method-overloading-breaks-all-kinds-of-inference-8).

Granted, I've been programming mostly in YAML over the past couple years, but that statement doesn't match my experience so far.